src/libcore/intrinsics.rs

```diff
@@ -1946,6 +1946,22 @@ extern "rust-intrinsic" {
     pub fn miri_start_panic(payload: *mut u8) -> !;
 }
 
+/// Defines the `count_code_region` intrinsic as a `LangItem`. `LangItem`s require a function body
+/// to register its DefId with the LangItem entry. The function body is never actually called (and
+/// is therefore implemented as an aborting stub) because it is replaced with the LLVM intrinsic
+/// `llvm.instrprof.increment` by
+/// `rustc_codegen_llvm::intrinsic::IntrinsicCallMethods::codegen_intrinsic_call()`.
+#[cfg(not(bootstrap))]
+#[cfg_attr(not(bootstrap), lang = "count_code_region")]
+fn count_code_region(_index: u32) {
```

Comment:
Is it possible to make this a regular intrinsic? This would allow you to remove `InstanceDef::InjectedCode`.

Reply:
I tried that at first but `InstanceDef::Intrinsic` implies a function signature with an `abi()` of `Abi::RustIntrinsic` or `Abi::PlatformIntrinsic`, which didn't fit with the implementation of the lang_item. Therefore it was more natural to create a new `InstanceDef` variant.

Reply:
If your question is, why not make this a regular intrinsic in the `extern` block, in `libcore::intrinsics`, the short version of the long story is, I needed a DefId for the placeholder function injected in the MIR pass. The `lang_item` seems to be a good solution to that problem, but regardless, a DefId requires an actual definition, and `extern` functions aren't "defined" (that is, assigned a DefId) until they are called. Since the code for this call is injected very late in the process (MIR passes), I couldn't use the `extern` by itself, but I could get a DefId by defining the function with a body. This worked much better, but it cannot be handled exactly the same as a regular intrinsic.

Reply:
I don't think the ABI of the Rust function really matters here. You're manually constructing a call to the LLVM `instrprof_increment` intrinsic during codegen, which means that the Rust `count_code_region` function will not exist in the output IR at all.

I think the only ABI that actually matters here is the ABI of the underling LLVM intrinsic, which I don't think needs to be explicitly set.

Reply:
>  the short version of the long story is, I needed a DefId for the placeholder function injected in the MIR pass.

You can use `#[lang]` on an extern item, for example:

https://github.com/rust-lang/rust/blob/219380d840877de97b0e17674be7212ebc929d51/src/libcore/panicking.rs#L93-L96

Reply:
Yes but only if it's called.

Reply:
All items in Rust end with a `DefId`, regardless of whether or not they are called. Did you run into an issue when trying to get the `DefId` of an extern item?

Reply:
Yes, I did. As I understand it (and witnessed), when compiling a crate that does not call an extern, the extern is never pulled into the AST and therefore never lowered to HIR. There is not an "Item" for that extern function to resolve, so a DefId is never assigned. And so the DefId is never added to the lang_items table (if the lang item is defined on an extern fn).

For an extern that *is* called in the crate that's being compiled, it does get a DefId and that DefId is then associated with the lang_item, but for what I'm calling "InjectedCode", this is not going to be the case.

I did try many permutations with the intrinsic definition, as an extern, not an extern, with a function that calls the extern, etc. The final solution was the cleanest working solution I could come up with.

Reply:
> Yes, I did. As I understand it (and witnessed), when compiling a crate that does not call an extern, the extern is never pulled into the AST and therefore never lowered to HIR. 

The AST is created long before we determine which methods are actually called. Even if we could determine that a method was never called during AST construction, it would still need a `DefId` so that it can be be resolved by downstream crates.

I tried out your branch locally - I think I found the problem you ran into. The `lang_items` code emits some fairly confusing error messages.

When you place a `#[lang]` attribute on an `extern` item, it gets treated as a 'weak lang item' by `src/librustc_passes/weak_lang_items.rs`. However, the 'weak lang item' system is intended for lang items which eventually get linked to a symbol from some other crate - which is not how you're using `count_code_region`.

I would recommend adding a special case for your lang item in `Context::register`:

https://github.com/rust-lang/rust/blob/219380d840877de97b0e17674be7212ebc929d51/src/librustc_passes/weak_lang_items.rs#L72-L83

You'll want to skip all of the logic in that method for `count_code_region` - although it's a foreign (`extern`) item, it's not a "weak lang item" as understood by that code.

That should cause `count_code_region` to be treated like a normal lang item, allowing you to obtain its `DefId` and insert a call to it into the MIR.

Reply:
Makes sense. I'll try it.

Reply:
Nice! That did the trick. Thanks for the recommendation.